File: Top_100_Liked_Questions/2._Add_Two_Numbers/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        curr = dummyHead = ListNode()
        carry = 0
        
        while l1 or l2 or carry:
            l1val = l1.val if l1 else 0
            l2val = l2.val if l2 else 0
            newVal = l1val + l2val + carry
            
            carry = newVal // 10
            curr.next = ListNode(newVal % 10)
            curr = curr.next
            
            if l1:
                l1 = l1.next
            
            if l2:
                l2 = l2.next
        
        return dummyHead.next

File: Top_100_Liked_Questions/2._Add_Two_Numbers/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode();
        ListNode curr = dummyHead;

        int carry = 0;
        while (l1 != null || l2 != null) {
            int a = l1 != null ? l1.val : 0;
            int b = l2 != null ? l2.val : 0;
            int sum = a + b + carry;

            curr.next = new ListNode(sum % 10);
            carry = Math.floorDiv(sum, 10);

            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
            curr = curr.next;
        }

        if (carry == 1)
            curr.next = new ListNode(1);

        return dummyHead.next;
    }
}